File: C#/proportional-simple-moving-average.cs
namespace QuantConnect.Algorithm.CSharp
{
    public class ProportionalSimpleMovingAverage : QCAlgorithm
    {
        private static Symbol _spy = QuantConnect.Symbol.Create("SPY", SecurityType.Equity, Market.USA);
		private static Symbol _qqq = QuantConnect.Symbol.Create("QQQ", SecurityType.Equity, Market.USA);
		private static Symbol _tlt = QuantConnect.Symbol.Create("TLT", SecurityType.Equity, Market.USA);
		private static Symbol _agg = QuantConnect.Symbol.Create("AGG", SecurityType.Equity, Market.USA);
		
		private Symbol _benchmark = _spy;
		
		private List<Symbol> _risk_on_symbols = new List<Symbol>{
			_spy,
			_qqq
		};
		private List<Symbol> _risk_off_symbols = new List<Symbol>{
			_tlt,
			_agg
		};
		
		private RollingWindow<decimal> _close_window;

        public override void Initialize()
        {
            SetStartDate(2016, 01, 01);  //Set Start Date
            SetEndDate(2016, 10, 14);    //Set End Date
            SetCash(10000);             //Set Strategy Cash

            // Find more symbols here: http://quantconnect.com/data
            // Forex, CFD, Equities Resolutions: Tick, Second, Minute, Hour, Daily.
            // Futures Resolution: Tick, Second, Minute
            // Options Resolution: Minute Only.
            AddEquity(_spy, Resolution.Daily);
            AddEquity(_qqq, Resolution.Daily);
            AddEquity(_tlt, Resolution.Daily);
            AddEquity(_agg, Resolution.Daily);

            //Build window
            _close_window = new RollingWindow<decimal>(84);
            IEnumerable<TradeBar> slices = History(_benchmark, 84);
            foreach(TradeBar bar in slices){
            	_close_window.Add(bar.Close);
            }
            
            Schedule.On(DateRules.EveryDay(_benchmark), 
            			TimeRules.AfterMarketOpen(_benchmark, 10),
            			EveryDayOnMarketOpen);
            
        }

        public void EveryDayOnMarketOpen(){
        	if (Transactions.GetOpenOrders().Count > 0){
        		return;
        	}
        	
        	IEnumerable<TradeBar> slices = History(_benchmark, 1);
        	TradeBar last_bar = slices.Last();
        	decimal bench_close = last_bar.Close;
        	
        	_close_window.Add(bench_close);
        	
        	decimal bench_mean_short = GetRollingAverage(21, _close_window);
        	decimal bench_mean_long = GetRollingAverage(84, _close_window);
        	
        	decimal risk_on_pct = (bench_mean_short / bench_close) *
        							((bench_mean_short * 2m / bench_mean_long) * .25m) /
        							_risk_on_symbols.Count;
        	decimal risk_off_pct = (bench_close / bench_mean_short) * 
        							((bench_mean_long * 2m / bench_mean_short) * .25m) /
        							_risk_off_symbols.Count;
        	
        	foreach (Symbol sid in _risk_on_symbols){
        		SetHoldings(sid, risk_on_pct);
        	}
        	foreach (Symbol sid in _risk_off_symbols){
        		SetHoldings(sid, risk_off_pct);
        	}
        }
        
        private decimal GetRollingAverage(int n, RollingWindow<decimal> window){
        	decimal sum = 0;
        	for (int i = 0; i < n; i++){
        		sum += window[i];
        	}
        	
        	return sum / n;
        }
    }
}

File: python/proportional-simple-moving-average.py
class ProportionalSMAFast(QCAlgorithm):

    def Initialize(self):
        # Set the cash we'd like to use for our backtest
        # This is ignored in live trading 
        self.SetCash(10000)
        
        # Start and end dates for the backtest.
        # These are ignored in live trading.
        self.SetStartDate(2016,01,01)
        self.SetEndDate(2016,10,14)
        
        # Add assets you'd like to see
        self.spy = self.AddEquity("SPY", Resolution.Daily).Symbol
        self.qqq = self.AddEquity("QQQ", Resolution.Daily).Symbol
        self.tlt = self.AddEquity("TLT", Resolution.Daily).Symbol
        self.agg = self.AddEquity("AGG", Resolution.Daily).Symbol
        
        self.benchmark = self.spy
        
        #Custom member variables
        self.risk_on_symbols = [self.spy, self.qqq]
        self.risk_off_symbols = [self.tlt, self.agg]
        
        #Schedule every day SPY is trading
        self.Schedule.On(self.DateRules.EveryDay(), \
                         self.TimeRules.AfterMarketOpen(self.benchmark, 10), \
                         Action(self.EveryDayOnMarketOpen))
    
    def EveryDayOnMarketOpen(self):
        #self.Log("Market Open!")
        #Do nothing if outstanding orders exist
        if self.Transactions.GetOpenOrders():
            return
        
        #Lookup last 84 days
        slices = self.History(self.benchmark, 84)
        #Get close of last (yesterday's) slice
        bench_close = slices["close"][-1]
        
        #Get mean over last 21 days
        bench_prices_short = slices["close"][-21:]
        bench_mean_short = bench_prices_short.mean()
        
        #Get mean over last 84 days
        bench_prices_long = slices["close"]
        bench_mean_long = bench_prices_long.mean()
        
        risk_on_pct  = (bench_mean_short/bench_close) * \
                        ((bench_mean_short *2 / bench_mean_long) *.25) / \
                        len(self.risk_on_symbols)
        risk_off_pct = (bench_close/bench_mean_short) * \
                        ((bench_mean_long *2 / bench_mean_short) *.25) / \
                        len(self.risk_off_symbols)
        
        #Submit orders                
        for sid in self.risk_on_symbols:
            self.SetHoldings(sid, risk_on_pct)
        for sid in self.risk_off_symbols:
            self.SetHoldings(sid, risk_off_pct)
